### ISSUE TITLE
Rewriting the GIT history of the upm tag/branch is causing issues for unity projects using Naugty Attributes

### ISSUE THREAD

**Issue by debox-dev** [dbrizov/NaughtyAttributes#188]
Hi,

Unity exclaims: "Fatal: reference is not a tree: [hash that unity locked on here]"

When I tried looking for the reference myself I couldn't find it as well.

We are using N.A and we love it, but it already happened twice that a commit we used was missing (due to history rewrite?) and suddenly people are having trouble opening the project because of the missing classes of N.A.

If this is the case, is it possible to be more subtle when considering history rewrites on the branches suggested for use with UPM?

**Comment by dbrizov** [comment]
The upm branch is deleted and recreated every time I push into master, it's automated. It's been like that since the branch was initially created. I don't know what I am supposed to do here. Also, how is this happening? I have personal projects that use NA, and haven't had such an issue when I release a new version.

**Comment by debox-dev** [comment]
Interesting.
Take a look at this commit: https://github.com/dbrizov/NaughtyAttributes/commit/ed0edab05e0503622f0b6c6b41061d5fb64e3ece
Obviously it exists, but it is no longer in the UPM branch.

Our unity project is locked on this commit from a past checkout of the UPM branch on our latest N.A upgrade

![image](https://user-images.githubusercontent.com/61346241/97861168-0a075d80-1d0c-11eb-8350-ecff3e8cd474.png)

How does this happen?

**Comment by dbrizov** [comment]
If you import NA via OpenUPM, this shouldn't happen, because it will create a scoped registry.
![image](https://user-images.githubusercontent.com/3619573/97864438-425d6a80-1d11-11eb-94d6-5ee2e10d4015.png)
You can also import it through the asset store.

I have to investigate how this hash is generated, and test. Also, you are right, this commit where the Samples folder is renamed is not part of the upm branch? That's pretty weird. I am not sure if it was like this before. I need to see why a commit created through the CI is not part of any branches. The thing is, I never had issues like this before, maybe Unity changed the way git packages are imported?

**Comment by dbrizov** [comment]
I will create a second **upm2** branch and test how I can automate it without deleting the history

**Comment by debox-dev** [comment]
@dbrizov  First thanks for the prompt response

Please consider you did not have any issues because you either upgrade the package often or because of the local unity cache.

We'll try the scoped approach, though we installed the package via the guidelines in the README.md here.

**Comment by dbrizov** [comment]
@debox-dev

Ok, I've never experienced this, because I've imported the package via **openupm**. I made some tests with another **upm2** branch, and I experienced this issue.

The CI is deleting the upm branch every time, exactly because it creates a new commit that is not part of any branch. This happens because the upm branch is a subtree branch. When I try to upgrade the subtree from the master branch without deleting it first - it fails because the subtree added another commit by itself, and they are no longer in sync (it's basically not a subtree anymore).

I'll just remove the **fix: Samples => Samples~** commit, and the history will remain, because I will be able to update the subtree branch without recreating it every time.

Unfortunately, I will have to regenerate the upm branch one more time, so you will experience this issue one last time.

The other thing I observed is that that the hash in the **packages-lock.json** acts like a version, and it's not updated automatically. If I update the upm branch, you will have to reimport it to get the new version. At least the old hash will still be in the history, and the project won't break.
